pkg/reconciler/awssnssource/subscription.go

```diff
@@ -0,0 +1,263 @@
+func (r *Reconciler) ensureSubscribed(ctx context.Context) error {
+	if skip.Skip(ctx) {
+		return nil
+	}
+
+	src := v1alpha1.SourceFromContext(ctx)
+	status := &src.(*v1alpha1.AWSSNSSource).Status
+
+	adapter, err := r.base.FindAdapter(src)
+	switch {
+	case apierrors.IsNotFound(err):
+		return nil
+	case err != nil:
+		return fmt.Errorf("finding receive adapter: %w", err)
+	}
+
+	url := adapter.Status.URL
+
+	// skip this cycle if the adapter URL wasn't yet determined
+	if !adapter.IsReady() || url == nil {
+		status.MarkNotSubscribed(v1alpha1.AWSSNSReasonNoURL,
+			"The receive adapter did not report its public URL yet")
+		return nil
+	}
+
+	spec := src.(apis.HasSpec).GetUntypedSpec().(v1alpha1.AWSSNSSourceSpec)
+
+	snsClient, err := newSNSClient(r.secretsCli(src.GetNamespace()), spec.ARN.Region, &spec.Credentials)
+	if err != nil {
+		status.MarkNotSubscribed(v1alpha1.AWSSNSReasonNoClient, "Cannot obtain SNS client")
+		return fmt.Errorf("%w", reconciler.NewEvent(corev1.EventTypeWarning, ReasonFailedSubscribe,
+			"Error creating SNS client: %s", err))
+	}
+
+	resp, err := snsClient.SubscribeWithContext(ctx, &sns.SubscribeInput{
+		Endpoint:              aws.String(url.String()),
+		Protocol:              &url.Scheme,
+		TopicArn:              aws.String(spec.ARN.String()),
+		Attributes:            spec.SubscriptionAttributes,
+		ReturnSubscriptionArn: aws.Bool(true),
+	})
+
+	switch {
+	case isAWSError(err):
+		// All documented API errors require some user intervention and
+		// are not to be retried.
+		// https://docs.aws.amazon.com/sns/latest/api/API_Subscribe.html#API_Subscribe_Errors
+		status.MarkNotSubscribed(v1alpha1.AWSSNSReasonRejected, "Subscription request rejected")
+		return controller.NewPermanentError(susbscribeErrorEvent(url, spec.ARN.String(), err))
+	case err != nil:
+		status.MarkNotSubscribed(v1alpha1.AWSSNSReasonFailedSync, "Cannot subscribe event source endpoint")
+		return fmt.Errorf("%w", susbscribeErrorEvent(url, spec.ARN.String(), err))
+	}
+
+	logging.FromContext(ctx).Debug("Subscribe responded with: ", resp)
+
+	status.MarkSubscribed()
+	status.SubscriptionARN = resp.SubscriptionArn
+
+	return reconciler.NewEvent(corev1.EventTypeNormal, ReasonSubscribed,
+		"Subscribed to SNS topic %q", spec.ARN.String())
+}
+
+// ensureUnsubscribed ensures the source's HTTP(S) endpoint is unsubscribed
+// from the SNS topic.
+func (r *Reconciler) ensureUnsubscribed(ctx context.Context) error {
+	src := v1alpha1.SourceFromContext(ctx)
+	// TODO(antoineco): Follow up with a proper FindSubscription() method.
+	// triggermesh/aws-event-sources#185
+	subsARN := src.(*v1alpha1.AWSSNSSource).Status.SubscriptionARN
+
+	// abandon if the subscription's ARN was never written to the source's status
+	if subsARN == nil {
+		return nil
+	}
+
+	spec := src.(apis.HasSpec).GetUntypedSpec().(v1alpha1.AWSSNSSourceSpec)
+
+	snsClient, err := newSNSClient(r.secretsCli(src.GetNamespace()), spec.ARN.Region, &spec.Credentials)
+	if err != nil {
```

Comment:
In case the user deletes the secret containing the AWS security credentials _before_ deleting the source itself, the finalizer will fail forever on the error
```
secrets "awscreds" not found
```
preventing the object from being deleted.

We need to check the returned error here, and give up if that error is unrecoverable.

Reply:
Done